PurchasesCoreSwift/Identity/PurchaserInfoManager.swift

```diff
@@ -0,0 +1,191 @@
+                if error != nil {
+                    self.deviceCache.clearPurchaserInfoCacheTimestamp(appUserID: appUserID)
+                } else if let info = purchaserInfo {
+                    self.cache(purchaserInfo: info, appUserID: appUserID)
+                }
+
+                if let completion = completion {
+                    self.operationDispatcher.dispatchOnMainThread {
+                        completion(purchaserInfo, error)
+                    }
+                }
+
+            }
+        }
+    }
+
+    @objc public func fetchAndCachePurchaserInfoIfStale(appUserID: String,
+                                                        isAppBackgrounded: Bool,
+                                                        completion: ReceivePurchaserInfoBlock?) {
+        let cachedPurchaserInfo = cachedPurchaserInfo(appUserID: appUserID)
+        let isCacheStale = deviceCache.isPurchaserInfoCacheStale(appUserID: appUserID,
+                                                                 isAppBackgrounded: isAppBackgrounded)
+        let needsToRefresh = isCacheStale || cachedPurchaserInfo == nil
+        if needsToRefresh {
+            Logger.debug(isAppBackgrounded
+                            ? Strings.purchaserInfo.purchaserinfo_stale_updating_in_background
+                            : Strings.purchaserInfo.purchaserinfo_stale_updating_in_foreground)
+            fetchAndCachePurchaserInfo(appUserID: appUserID,
+                                       isAppBackgrounded: isAppBackgrounded,
+                                       completion: completion)
+            Logger.rcSuccess(Strings.purchaserInfo.purchaserinfo_updated_from_network)
+        } else {
+            if let completion = completion {
+                operationDispatcher.dispatchOnMainThread {
+                    completion(cachedPurchaserInfo, nil)
+                }
+            }
+        }
+    }
+
+    @objc(sendCachedPurchaserInfoIfAvailableForAppUserID:)
+    public func sendCachedPurchaserInfoIfAvailable(appUserID: String) {
+        guard let info = cachedPurchaserInfo(appUserID: appUserID) else {
+            return
+        }
+
+        sendUpdateIfChanged(purchaserInfo: info)
+    }
+
+    @objc(purchaserInfoWithAppUserID:completionBlock:)
+    public func purchaserInfo(appUserID: String, completionBlock: ReceivePurchaserInfoBlock?) {
+        let infoFromCache = cachedPurchaserInfo(appUserID: appUserID)
+        var completionCalled = false
+
+        if let infoFromCache = infoFromCache {
+            Logger.debug(Strings.purchaserInfo.vending_cache)
+            if let completion = completionBlock {
+                completionCalled = true
+                operationDispatcher.dispatchOnMainThread {
+                    completion(infoFromCache, nil)
+                }
+            }
+        }
+
+        // Prevent calling completion twice.
+        let fetchPurchaserInfoCompletion = completionCalled ? nil : completionBlock
```

Comment:
I'd love to reflect that in the variable name, not sure what's the best way to do it, i.e.: `maybeCompletion`, `completionIfNotCalledAlready`

Reply:
I like `completionIfNotCalledAlready`
it's kinda weird, but like, it works.